satellite/console/consoleweb/consoleapi/referrals.go

```diff
@@ -0,0 +1,216 @@
+		controller.log.Error("token handler could not encode token response", zap.Error(ErrReferralsAPI.Wrap(err)))
+		return
+	}
+}
+
+// Register creates new user, sends activation e-mail.
+func (controller *Referrals) Register(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	var err error
+	defer mon.Task()(&ctx)(&err)
+
+	var registerData struct {
+		FullName      string `json:"fullName"`
+		ShortName     string `json:"shortName"`
+		Email         string `json:"email"`
+		PartnerID     string `json:"partnerId"`
+		Password      string `json:"password"`
+		ReferralToken string `json:"referralToken"`
+	}
+
+	err = json.NewDecoder(r.Body).Decode(&registerData)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	userID, err := uuid.New()
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	err = controller.referralsService.ReferralManagerConn(ctx)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	referralToken, err := uuid.Parse(registerData.ReferralToken)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	err = controller.referralsService.RedeemToken(ctx, userID, referralToken)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	// need to generate a registration token for the referred user?
+	// TODO: add user ID onto CreateUser
+	user, err := controller.service.CreateUser(ctx,
+		console.CreateUser{
+			FullName:  registerData.FullName,
+			ShortName: registerData.ShortName,
+			Email:     registerData.Email,
+			PartnerID: registerData.PartnerID,
+			Password:  registerData.Password,
+		},
+		console.RegistrationSecret{},
+		"",
+	)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	token, err := controller.service.GenerateActivationToken(ctx, user.ID, user.Email)
+	if err != nil {
+		controller.serveJSONError(w, err)
+		return
+	}
+
+	link := controller.ExternalAddress + "activation/?token=" + token
+	userName := user.ShortName
+	if user.ShortName == "" {
+		userName = user.FullName
+	}
+
+	controller.mailService.SendRenderedAsync(
+		ctx,
+		[]post.Address{{Address: user.Email, Name: userName}},
+		&consoleql.AccountActivationEmail{
+			ActivationLink: link,
+			Origin:         controller.ExternalAddress,
+		},
+	)
+
+	w.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(w).Encode(user.ID)
+	if err != nil {
+		controller.log.Error("registration handler could not encode userID", zap.Error(ErrReferralsAPI.Wrap(err)))
+		return
+	}
+}
+
+// serveJSONError writes JSON error to response output stream.
+func (controller *Referrals) serveJSONError(w http.ResponseWriter, err error) {
+	w.WriteHeader(controller.getStatusCode(err))
+
+	var response struct {
+		Error string `json:"error"`
+	}
+
+	response.Error = err.Error()
+
+	err = json.NewEncoder(w).Encode(response)
+	if err != nil {
+		controller.log.Error("failed to write json error response", zap.Error(ErrAuthAPI.Wrap(err)))
```

Comment:
controller.log.Error("failed to write json error response", zap.Error(ErrReferralsAPI.Wrap(err)))